agent/utils/buffer-ring/bigqueue/bigqueue.go

```diff
@@ -0,0 +1,208 @@
+		if size := proto.Size(msg); size < 100 {
+			r.l.Debugf("inserting message (%d bytes): %s.", size, &out)
+		} else {
+			r.l.Debugf("inserting message (%d bytes):\n%s\n", size, prototext.Format(msg))
+		}
+	}
+	size := uint32(len(out))
+	if size > r.totalSize {
+		r.l.Errorf("data size: '%d' overflows free cache space: '%d'", size, r.totalSize)
+		return
+	}
+	if newOffset := size + atomic.LoadUint32(&r.sizeInUse); newOffset > r.totalSize {
+		if err := r.drain(newOffset - r.totalSize + size); err != nil {
+			r.l.Errorf("draining cache: %+v", err)
+			return
+		}
+	}
+	r.m.Lock()
+	_, err = r.fq.Enqueue(out)
+	r.m.Unlock()
+	if err != nil {
+		r.l.Errorf("inserting to cache: %+v", err)
+		return
+	}
+	atomic.AddUint32(&r.sizeInUse, size)
+}
+
+// Send unmarshals agent message and send it in loop until dries the cache.
+func (r *Ring) Send(send func(msg *agentpb.AgentMessage) error) {
+	for {
+		r.m.Lock()
+		idx, b, err := r.fq.Dequeue()
+		r.m.Unlock()
+		if err != nil {
+			r.l.Errorf("reading entry from cache: %+v", err)
+			return
+		}
+		if idx == -1 {
+			return
+		}
+		atomic.AddUint32(&r.sizeInUse, ^uint32(len(b)-1))
+		var out agentpb.AgentMessage
+		if err := proto.Unmarshal(b, &out); err != nil {
+			r.l.Errorf("unmarshal entry from cache: %+v", err)
+			return
+		}
+		if err := send(&out); err != nil {
+			r.l.Errorf("sending entry from cache: %+v", err)
+			return
+		}
+	}
+}
+
+// Close closes cache database.
+func (r *Ring) Close() error {
+	select {
+	case <-r.done:
+	default:
+		close(r.done)
+	}
+	return r.fq.Close()
+}
+
+// initPaths creates all paths for queue to use. Original repo creates directories with perm error.
+func initPaths(dir string) error {
+	for _, path := range []string{
+		"",
+		bigqueue.IndexFileName,
+		bigqueue.DataFileName,
+		bigqueue.MetaFileName,
+		bigqueue.FrontFileName,
+	} {
+		if err := os.MkdirAll(filepath.Join(dir, path), os.ModePerm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// fillInUse fills sizeInUse that represents the amount of space used in bigqueue with messages data.
+func (r *Ring) fillInUse() error {
+	for i := uint64(0); ; i += paginationSize {
+		data, _, err := r.fq.PeekPagination(i, paginationSize)
+		if err != nil {
+			return err
+		}
+		if len(data) == 0 {
+			return nil
+		}
+		for j := range data {
+			r.sizeInUse += uint32(len(data[j]))
+		}
+	}
+}
+
+// drain tries to drain cache.
+func (r *Ring) drain(amount uint32) error {
+	for size := uint32(0); size < amount; {
+		r.m.Lock()
+		idx, b, err := r.fq.Dequeue()
+		r.m.Unlock()
+		if err != nil {
+			return err
+		}
+		if idx == -1 {
+			return nil
+		}
+		size += uint32(len(b))
+		atomic.AddUint32(&r.sizeInUse, ^uint32(len(b)-1))
+	}
+	return nil
+}
+
+// autoGC delete all used data files and free disk space.
+func (r *Ring) autoGC() {
+	go func() {
+		ticker := time.NewTicker(time.Second * AutoGCBySeconds)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				r.m.Lock()
```

Comment:
Let's also drain the cache here in case the size is bigger than the limit or close to it.